### ISSUE TITLE
[Feature]:  Review logits post-processing performance

### ISSUE THREAD

**Issue by nzmora-nvidia** [NVIDIA/TensorRT-LLM#7532]
### 🚀 The feature, motivation and pitch

@lucaslie found differences between AD's and PT's logits post-processing implementations and we should study these to understand the perf implications of the 2 implementations.

Slack: https://nvidia.slack.com/archives/C08T55LHSG4/p1756943076165769

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

**Comment by lucaslie** [comment]
One important optimization would be to compute the output of lm_head for the last token only if we don't need all the logits. To compute the logits for all tokens with very long ISL can be quite slow, see https://nvidia.slack.com/archives/C090ARR8U6Q/p1764213657177919?thread_ts=1764060128.823639&cid=C090ARR8U6Q.

cc @nvpohanh

**Comment by lucaslie** [comment]
Here is a quick diff @nvchenghaoz and I discussed to bring logits_gather_ids into the AutoDeploy:

```
diff --git a/tensorrt_llm/_torch/auto_deploy/config/default.yaml b/tensorrt_llm/_torch/auto_deploy/config/default.yaml
index 42d4e763d9..e8ce7b3c5a 100644
--- a/tensorrt_llm/_torch/auto_deploy/config/default.yaml
+++ b/tensorrt_llm/_torch/auto_deploy/config/default.yaml
@@ -171,6 +171,10 @@ transforms:
   multi_stream_moe:
     stage: compile
     enabled: false
+  gather_logits_before_lm_head:
+    # do a gather before lm_head with hidden_states[logit_gather_ids[:total_num_tokens]]
+    stage: compile
+    enabled: true
   compile_model:
     stage: compile
     run_per_gm: false
diff --git a/tensorrt_llm/_torch/auto_deploy/custom_ops/attention_interface.py b/tensorrt_llm/_torch/auto_deploy/custom_ops/attention_interface.py
index 818e560578..ce675e6e99 100644
--- a/tensorrt_llm/_torch/auto_deploy/custom_ops/attention_interface.py
+++ b/tensorrt_llm/_torch/auto_deploy/custom_ops/attention_interface.py
@@ -98,6 +98,8 @@ class SequenceInfo:
       with sequence 1 in the batch.
     - slot_idx: [s_0, s_1, ..., s_{b-1}]
       Corresponds to the slot index of each sequence in the batch.
+    - logit_gather_ids: [l_0, l_1, ..., l_{b-1}]
+      Corresponds to the index of the last token in the sequence that needs to be gathered for each sequence in the batch.
 
     ################################################################################################
 
@@ -201,6 +203,7 @@ class SequenceInfo:
             "cache_loc": torch.empty(max_num_cache_loc_assignments, dtype=torch.int),
             "pages_per_seq": torch.empty(self.max_batch_size, dtype=torch.int),
             "slot_idx": torch.empty(self.max_batch_size, dtype=torch.long),
+            "logit_gather_ids": torch.empty(self.max_num_tokens, dtype=torch.int),
             # OTHER FIELDS WHERE WE NEED EFFICIENT HOST<>DEVICE TRANSFER
             "_gather_idx": torch.empty(self.max_num_tokens, dtype=torch.int),
         }
@@ -209,7 +212,14 @@ class SequenceInfo:
         }
         # NOTE: order of keys is relevant here!
         self._uncached_arg_names = ("input_ids", "position_ids")
-        self._cached_arg_names = ("seq_len", "input_pos", "cache_loc", "pages_per_seq", "slot_idx")
+        self._cached_arg_names = (
+            "seq_len",
+            "input_pos",
+            "cache_loc",
+            "pages_per_seq",
+            "slot_idx",
+            "logit_gather_ids",
+        )
         # page_size is the size of attentionkv-cache pages.
         # chunk_size is used in mamba prefill kernels to split the context into chunks.
         self._cached_constants = ("page_size", "chunk_size")
@@ -638,6 +648,7 @@ class SequenceInfo:
         input_pos: Optional[Union[Sequence[int], int]] = None,
         page_assignments: Optional[Sequence[Sequence[int]]] = None,
         slot_idx: Optional[Sequence[int]] = None,
+        logit_gather_ids: Optional[Sequence[int]] = None,
         **extra_args: Dict[str, Union[torch.Tensor, Sequence[torch.Tensor]]],
     ) -> None:
         """Create and store sequence information for the next forward pass.
@@ -648,6 +659,7 @@ class SequenceInfo:
             input_pos: Absolute starting position in the cache for each sequence.
             page_assignments: List of sequences of page assignments for each sequence.
             slot_idx: List of slot indices for each sequence.
+            logit_gather_ids: List of logit gather indices for each sequence.
             extra_args: Extra arguments to be stored in the interface.
 
         This i/f will ensure that all sequence info args are updated accordingly. Reset values are
@@ -683,6 +695,12 @@ class SequenceInfo:
             free_slot_idx = self._get_unique_value(set(slot_idx), self.max_batch_size)
             self._store_arg("slot_idx", slot_idx, reset_val=free_slot_idx)
 
+        # check for updated logit_gather_ids
+        if logit_gather_ids is not None:
+            self._store_arg("logit_gather_ids", logit_gather_ids, reset_val=0)
+        else:
+            self._store_arg("logit_gather_ids", list(range(self.total_num_tokens)))
+
         ### UPDATE MAIN INPUTS #####################################################################
         # set new input_ids and make sure to flatten it
         self._store_arg("input_ids", self._flatten(input_ids))
diff --git a/tensorrt_llm/_torch/auto_deploy/shim/ad_executor.py b/tensorrt_llm/_torch/auto_deploy/shim/ad_executor.py
index d74162de47..17d258b8ac 100644
--- a/tensorrt_llm/_torch/auto_deploy/shim/ad_executor.py
+++ b/tensorrt_llm/_torch/auto_deploy/shim/ad_executor.py
@@ -206,6 +206,7 @@ class ADEngine(ModelEngine):
         input_ids: List[List[int]] = []
         input_pos: List[int] = []
         last_logit_only: List[bool] = []
+        logit_gather_ids: List[int] = []
         page_assignments: List[List[int]] = []
         slot_idx: List[int] = []
         flat_gather_idx: List[int] = []
@@ -232,6 +233,7 @@ class ADEngine(ModelEngine):
 
             request.py_batch_idx = request.seq_slot
             last_logit_only.append(True)
+            logit_gather_ids.append(num_ctx_tokens - 1)
 
             # get cache indices and truncate the number of blocks according to end_compute
             cache_indices = kv_cache_manager.get_cache_indices(request)
@@ -267,6 +269,7 @@ class ADEngine(ModelEngine):
 
             # return all logits
             last_logit_only.append(False)
+            logit_gather_ids.append(num_ctx_tokens + num_generation_tokens - 1)
 
             # get cache indices
             cache_indices = kv_cache_manager.get_cache_indices(request)
@@ -278,6 +281,7 @@ class ADEngine(ModelEngine):
             input_pos=input_pos,
             page_assignments=page_assignments,
             slot_idx=slot_idx,
+            logit_gather_ids=logit_gather_ids,
             **extra_args,
         )
         # scatter the new tokens into the input_ids tensor if provided
@@ -317,6 +321,7 @@ class ADEngine(ModelEngine):
     ):
         """Run forward from scheduled requests; main entrypoint that gets called by the executor."""
         # convert requests and store in sequence info object
+        assert not gather_context_logits, "gather_context_logits is not supported yet"
         new_tokens = getattr(new_tensors_device, "new_tokens", None)
         last_logit_only = self._prepare_inputs(scheduled_requests, resource_manager, new_tokens)
         self.iter_counter += 1
```
